parlai/crowdsourcing/utils/README.md

```diff
@@ -14,3 +14,54 @@
 - Add the code for this in `.check_messages()`, inside a `if 'check_name' in violation_types:` condition
 - Add the name of the check to `self.ALL_VIOLATION_TYPES`; otherwise, this check will not be recognized, and an error will be raised if the check is specified when calling `.check_messages()`!
 - To use the check: add the name of the check to the `violation_types` arg when calling `.check_messages()`
+
+## `analysis.py`
+
+Contains abstract classes that provide the basic functionalities for compiling data from a Mephsito task.
+Mephisto provides two interfaces for retrieving its crowdsourced data; `MephistoDB` and `DataBrowser`.
+Using `AbstractResultsCompiler` you do not need to directly interact with these two modules---it provides an abstraction on top of these two.
+This class has methods such as `get_task_data` and `get_task_units` which handles interacting with Mephisto abstractions.
+For compiling your dataset from your crowdsourced Mephisto task, you need to extend this class and implement the following methods:
+
+* `compile_results` that returns a python *dictionary* (key-value pairs) or a pandas *dataframe*. We assume that, each unit of the crowdsourcing task (for example, annotation or conversation) has a unique id. This id is the key in the former case (the data from the unit is the value) and the row id for the latter.
+
+* (optoinal) `is_unit_acceptable` helps with simple filtering and data clean up. It receives the data from a unit of work and returns a boolean. We discrad this unit if it returns `False`.
+
+### Example
+Imagine you have a Mephisto task that the output of each unit of its work looks like this:
+```.python
+{
+    'ID': 1234,
+    'favorite_flower': 'rose',
+    'favorite_season', 'winter',
+    'status': 'accepted'
+}
+```
+
+Let's say this task is called `flowers_crowdsourcing` and want to discard every participant with `status` being "accepted".
+Here is how you can have your data compiled and saved:
+
+```.python
+from parlai.crowdsourcing.utils.analysis import AbstractResultsCompiler
+
+class MyResultsCompiler(AbstractResultsCompiler):
+    def is_unit_acceptable(self, unit_data):
+        return unit_data['status'] ==  'accepted'
+
+    def compile_results(self):
+        data = dict()
+        for work_unit in self.get_task_data():
+            unit_id = work_unit.pop('ID)
```

Comment:
hmm do we want to mutate `work_unit` here by popping the ID?

Reply:
Yeah, but we are not reusing `work_unit` objects after this. We could keep the id but then it is our key for retrieving some content that includes the key itself. It's a bit of redundancy.

Reply:
Okay, fair, I don't have a strong preference here. But can you add the closing `'` after `ID`?